src/replication/packing_stream.hpp

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (C) 2008 Search Solution Corporation. All rights reserved by Search Solution.
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+/*
+ * packing_stream.hpp
+ */
+
+#ident "$Id$"
+
+#ifndef _PACKING_STREAM_HPP_
+#define _PACKING_STREAM_HPP_
+
+#include <vector>
+#include <functional>
+#include "stream_common.hpp"
+#include "storage_common.h"
+
+class buffer_provider;
+class packable_object;
+class object_builder;
+class stream_packer;
+class log_generator;
+
+typedef enum
+{
+  COLLECT_ALL_BUFFERS = 0,
+  COLLECT_ONLY_FILLED_BUFFERS
+} COLLECT_FILTER;
+
+typedef enum
+{
+  COLLECT_KEEP = 0,
+  COLLECT_AND_DETACH
+} COLLECT_ACTION;
+
+class stream_handler
+{
+public:
+  virtual int handling_action (BUFFER_UNIT *ptr, size_t byte_count) = 0;
+};
+
+class stream_entry : public pinner
+{
+private:
+  int stream_entry_id;
+
+  /* TODO : should move this to stream_packer ? */
+  buffer_context *m_buffered_range;
+
+  bool m_is_packable;
+
+protected:
+  std::vector <packable_object *> m_packable_entries;
+
+  packing_stream *m_stream;
+
+  stream_position m_data_start_position;
+
+  virtual object_builder *get_builder () = 0;
+
+public:
+  stream_entry (packing_stream *stream);
```

Comment:
you can add "= NULL" at the end of "packing_stream *stream = NULL" to provide a default value